### ISSUE TITLE
conditionally link against libatomic on non-clang builds

### ISSUE THREAD

**Issue by senior7515** [scylladb/seastar#657]
Change:

a44cedf287bfbe8c0c07ad663f26f0683e5b0885
Author:     Michael Munday <mike.munday@ibm.com>

Introduced `-latomic` 

```
When compiling functions in <atomic> the compiler may emit calls to
functions in libatomic if the hardware doesn't support the required
atomic operation.

This commit fixes the build on IBM Z (s390x) using the devtoolset-7
toolchain.
```
(related #530 )

These are not needed with Clang:

https://llvm.org/docs/Atomics.html

Change 

c9d6e20b18a07e54c9406ee6b62c1e1f5b339d65
Author:     Jesse Haber-Kucharsky <jhaberku@scylladb.com>

Introduced unconditionally linking against libatomic which is incorrect if the CC=clang, CXX=clang++ 

Will submit patch to conditionally link -latomic when using GCC

**Comment by espindola** [comment]
Does it cause any problem to have an extra -latomic in the link line?

**Comment by senior7515** [comment]
@espindola yes. We are trying to ship a tarball to a production server w/ all deps build at specific hashes/versions, etc for reproducibility, including compilers (also for regulations) 

So briefly,

1) We shouldn't be linking against things that aren't needed at runtime
2) To conditionally link against it is pretty easy fix
3) It is a problem with packaging as it causes us to add unnecessary complexity (if lib in whitelist; then skip)

**Comment by espindola** [comment]
OK. If the problem is just the runtime, adding -Wl,--as-needed is probably the best solution, as it is simple and will handle all libraries.

**Comment by senior7515** [comment]
The problem is unnecessary packaging complexity.

**Comment by senior7515** [comment]
Also, I'm not sure what the issue is adding a patch for conditionally linking w/ libatomic.

**Comment by espindola** [comment]
Complexity. I would be happy to approve a patch that adds -Wl,--as-needed, since it would be a simple change and it would work for all libraries.

The only cost is to have libatomic.so on the *build* machine, which seems trivial.
